File: src/net/dinkla/raytracer/brdf/BRDF.java
package net.dinkla.raytracer.brdf;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.math.Vector3DF;
import net.dinkla.raytracer.samplers.Sampler;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 12.04.2010
 * Time: 14:16:43
 * To change this template use File | Settings | File Templates.
 */
public abstract class BRDF<C extends Color> {

    public Sampler sampler;

    public class Sample {
        public C color;
        public Vector3DF wi;
        public float pdf;
    }

    abstract public C f(final Shade sr, final Vector3DF wo, final Vector3DF wi);

    abstract public Sample sampleF(Shade sr, Vector3DF wo);

    abstract public C rho(Shade sr, Vector3DF wo);
    
}

File: src/main/kotlin/net/dinkla/raytracer/brdf/BRDF.kt
package net.dinkla.raytracer.brdf

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.math.Vector3D

abstract class BRDF {

    class Sample(
            val wi: Vector3D,
            val color: Color,
            val pdf: Double
    )

    abstract fun f(sr: Shade, wo: Vector3D, wi: Vector3D): Color

    abstract fun sampleF(sr: Shade, wo: Vector3D): Sample

    abstract fun rho(sr: Shade, wo: Vector3D): Color
}
